### PR TITLE
Support IMDSv2 to retrieve metadata

### PR DESCRIPTION
**Pull request by y26805**
Fixes https://github.com/HENNGE/aiodynamo/issues/157

Tested on an IMDSv2-only instance on AWS EC2, and it works 🎉

### CODE REVIEW

**Comment by ojii** [comment]
shouldn't we support both flavors of IMDS?

**Comment by y26805** [comment]
> shouldn't we support both flavors of IMDS?

IMDSv2 is supposed to be more secure (so AWS will probably remove v1 support at some point). Also, it's currently not possible to only allow IMDSv2, based on AWS documentation.
https://github.com/HENNGE/aiodynamo/issues/157#issuecomment-1742510749

File: src/aiodynamo/credentials.py [hunk 1]
```diff
@@ -407,6 +408,32 @@ async def fetch_metadata(self, http: HttpImplementation) -> Metadata:
...
+                max_attempts=self.max_attempts,
+                timeout=self.timeout,
+                request=Request(
+                    method="PUT",
+                    url=str(token_url),
+                    headers={
+                        "X-aws-ec2-metadata-token-ttl-seconds": str(
+                            session_duration_seconds
+                        )
+                    },
```
Comment (ojii) [inline src/aiodynamo/credentials.py:427]: shouldn't we cache this token if it has a TTL?

File: src/aiodynamo/credentials.py [hunk 2]
```diff
@@ -354,9 +355,94 @@ async def fetch_metadata(self, http: HttpImplementation) -> Metadata:
 
 # https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html
 @dataclass
-class InstanceMetadataCredentials(MetadataCredentials):
+class InstanceMetadataCredentialsWithImdsV2(MetadataCredentials):
```
Comment (ojii) [inline src/aiodynamo/credentials.py:358]: suuuper nitpicky, but "InstanceMetadata" and "Imds" are redundant. I think `InstanceMetadataCredentialsV1` and `InstanceMetadataCredentialsV2` would be more appropriate

File: tests/unit/test_credentials.py [hunk 3]
```diff
@@ -129,15 +152,23 @@ async def test_ec2_instance_metdata_credentials(
...
-    http: HttpImplementation, instance_metadata_server: InstanceMetadataServer
+    http: HttpImplementation,
+    instance_metadata_server: InstanceMetadataServer,
+    use_v2: bool,
 ) -> None:
+    model = (
+        InstanceMetadataCredentialsWithImdsV2
+        if use_v2
+        else InstanceMetadataCredentialsWithImdsV1
+    )
```
Comment (ojii) [inline tests/unit/test_credentials.py:165]: why not parametrize model using `parametrize("model", [InstanceMetadataCredentialsWithImdsV2, InstanceMetadataCredentialsWithImdsV1])`?

File: tests/unit/test_credentials.py [hunk 4]
```diff
@@ -60,7 +65,9 @@ async def credentials_handler(self, request: web.Request) -> web.Response:
 
 
 @pytest.fixture
-async def instance_metadata_server() -> AsyncGenerator[InstanceMetadataServer, None]:
+async def instance_metadata_server(
+    use_v2: bool = True,
```
Comment (ojii) [inline tests/unit/test_credentials.py:69]: that's not how fixtures work... either you need to use marks, another fixture, or just always include v2

File: src/aiodynamo/credentials.py (hunk 2 above)
Comment (y26805) [reply src/aiodynamo/credentials.py:358]: yeah i was thinking how to name them.. `InstanceMetadataCredentialsV2`/V1 sounds good. will change

File: tests/unit/test_credentials.py (hunk 4 above)
Comment (y26805) [reply tests/unit/test_credentials.py]: i was thinking about having 2 fixtures, but it feels like an overkill... so yeah let me revert that

**Review by ojii — APPROVED** [review]

**Pull request merged**